### ISSUE TITLE
AutoConfigurations should gracefully do nothing if dependencies are missing

### ISSUE THREAD

**Issue by graemerocher** [spring-projects/spring-boot#4532]
Currently we have issues in Grails popping up when Boot's AutoConfiguration logic kicks in. The latest one is `org.springframework.boot.autoconfigure.mongo.MongoAutoConfiguration` which throws exceptions if the MongoDB driver is on the classpath but the Spring Data MongoDB classes are not.

There are many cases where you would want to use the MongoDB driver (for example with GORM) but not Spring Data MongoDB, but right now Boot just dies with:

```
Exception in thread "main" org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.boot.autoconfigure.mongo.MongoAutoConfiguration': Injection of autowired dependencies failed; nested exception is org.springframework.beans.factory.BeanCreationException: Could not autowire field: private org.springframework.boot.autoconfigure.mongo.MongoProperties org.springframework.boot.autoconfigure.mongo.MongoAutoConfiguration.properties; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'spring.data.mongodb.CONFIGURATION_PROPERTIES': Initialization of bean failed; nested exception is java.lang.TypeNotPresentException: Type org.springframework.data.mapping.model.FieldNamingStrategy not present
```

We had similar issues with the other AutoConfigurations, and the only way to fix this is to manually exclude them from your `Application` class. It seems to be it would be much better if Boot simply did not process the `AutoConfiguration` if the dependencies are not present. This will  lead to fewer mysterious errors and a more flexible system.

**Comment by wilkinsona** [comment]
>  It seems to be it would be much better if Boot simply did not process the AutoConfiguration if the dependencies are not present.

That's already the intention. If that's not happening then it sounds like we're missing one or more `@ConditionalOnClass` annotations.

> The latest one is org.springframework.boot.autoconfigure.mongo.MongoAutoConfiguration which throws exceptions if the MongoDB driver is on the classpath but the Spring Data MongoDB classes are not.

I just tried adding Mongo's driver to a an app without Spring Data MongoDB and it did the right thing. The pure Mongo auto-configuration kicked in, but the Spring Data MongoDB auto-configuration did not.

What version of Spring Boot are you using and what's the specific failure that you're seeing?

**Comment by wilkinsona** [comment]
It looks like this was broken sometime during 1.3.0's development and was fixed in 1.3.0.RC1. See https://github.com/spring-projects/spring-boot/issues/4049.

**Comment by graemerocher** [comment]
Ah ok, should have checked. Thanks for the pointer
